File: backend/app/services/stock_asset_class.py
from __future__ import annotations

import re
from typing import Optional

ASSET_CLASSES = ("Equity", "Debt", "Gold")

# NSE/BSE/Zerodha sovereign gold bond symbols, e.g. SGBFEB32, SGBFEB32IV-GB
SGB_SYMBOL_PATTERN = re.compile(r"^SGB", re.IGNORECASE)
# RBI sovereign gold bond ISIN prefix
SGB_ISIN_PREFIX = "IN0020"


def normalize_stock_symbol(symbol: str) -> str:
    value = symbol.strip().upper()
    if "-" in value:
        value = value.split("-", 1)[0]
    return value


def classify_stock(
    symbol: str,
    isin: Optional[str] = None,
    name: Optional[str] = None,
) -> str:
    normalized_symbol = normalize_stock_symbol(symbol)
    if SGB_SYMBOL_PATTERN.match(normalized_symbol):
        return "Gold"

    if isin:
        isin_upper = isin.strip().upper()
        if isin_upper.startswith(SGB_ISIN_PREFIX):
            return "Gold"

    if name:
        name_upper = name.upper()
        if "SOVEREIGN GOLD BOND" in name_upper or " GOLD BOND" in name_upper:
            return "Gold"

    return "Equity"

File: frontend/src/utils/stockAssetClass.js
import { normalizeAssetClass } from './categoryAllocation'

// Keep in sync with backend/app/services/stock_asset_class.py
const SGB_SYMBOL_PATTERN = /^SGB/i
const SGB_ISIN_PREFIX = 'IN0020'

export function normalizeStockSymbol(symbol) {
  const value = (symbol ?? '').trim().toUpperCase()
  const dashIndex = value.indexOf('-')
  return dashIndex === -1 ? value : value.slice(0, dashIndex)
}

export function classifyStockAssetClass({ symbol, isin, name, asset_class: assetClass } = {}) {
  const normalizedSymbol = normalizeStockSymbol(symbol)
  if (SGB_SYMBOL_PATTERN.test(normalizedSymbol)) {
    return 'Gold'
  }

  const isinUpper = (isin ?? '').trim().toUpperCase()
  if (isinUpper.startsWith(SGB_ISIN_PREFIX)) {
    return 'Gold'
  }

  const nameUpper = (name ?? '').toUpperCase()
  if (nameUpper.includes('SOVEREIGN GOLD BOND') || nameUpper.includes(' GOLD BOND')) {
    return 'Gold'
  }

  if (assetClass) {
    return normalizeAssetClass(assetClass)
  }

  return 'Equity'
}
